connectivity/drivers/lora/TARGET_STM32WL/STM32WL_LoRaRadio.cpp

```diff
@@ -0,0 +1,1561 @@
+                                  + ((_packet_params.params.gfsk.crc_length == RADIO_CRC_2_BYTES) ? 2.0f : 0.0f))
+                             / _mod_params.params.gfsk.bit_rate) * 1000);
+        }
+        break;
+        case MODEM_LORA: {
+            float ts = lora_symbol_time[_mod_params.params.lora.bandwidth - 4][12
+                                                                               - _mod_params.params.lora.spreading_factor];
+            // time of preamble
+            float t_preamble = (_packet_params.params.lora.preamble_length + 4.25f) * ts;
+            // Symbol length of payload and time
+            float tmp = ceil((8 * pkt_len - 4 * _mod_params.params.lora.spreading_factor
+                              + 28 + 16 * _packet_params.params.lora.crc_mode
+                              - ((_packet_params.params.lora.header_type == LORA_PACKET_FIXED_LENGTH) ? 20 : 0))
+                             / (float)(4 * (_mod_params.params.lora.spreading_factor
+                                            - ((_mod_params.params.lora.low_datarate_optimization > 0) ? 2 : 0))))
+                        * ((_mod_params.params.lora.coding_rate % 4) + 4);
+            float n_payload = 8 + ((tmp > 0) ? tmp : 0);
+            float t_payload = n_payload * ts;
+            // Time on air
+            float tOnAir = t_preamble + t_payload;
+            // return milliseconds (as ts is in milliseconds)
+            air_time = floor(tOnAir + 0.999);
+        }
+        break;
+    }
+
+    return air_time;
+}
+
+void STM32WL_LoRaRadio::radio_reset()
+{
+//    _reset_ctl.output();
+//    _reset_ctl = 0;
+//    // should be enough, required is 50-100 us
+//    ThisThread::sleep_for(2);
+//    _reset_ctl.input();
+    // give some time for automatic image calibration
+    ThisThread::sleep_for(6);
+}
+
+void STM32WL_LoRaRadio::wakeup()
+{
+    // hold the NSS low, this should wakeup the chip.
+    // now we should wait for the _busy line to go low
+    if (_operation_mode == MODE_SLEEP) {
+//        _chip_select = 0;
+//        wait_us(100);
+//        _chip_select = 1;
+//        //wait_us(100);
+#if MBED_CONF_STM32WL_LORA_DRIVER_SLEEP_MODE == 1
+        wait_us(3500);
+        // whenever we wakeup from Cold sleep state, we need to perform
+        // image calibration
+        _force_image_calibration = true;
+        cold_start_wakeup();
+#endif
+    }
+}
+
+void STM32WL_LoRaRadio::sleep(void)
+{
+
+
+
+#if MBED_CONF_STM32WL_LORA_DRIVER_SLEEP_MODE == 1
+    // cold start, power consumption 160 nA
+    sleep_state = 0x00;
+#endif
+  
+  /* switch the antenna OFF by SW */
+    RBI_ConfigRFSwitch(RBI_SWITCH_OFF);
+    Radio_SMPS_Set(SMPS_DRIVE_SETTING_DEFAULT);
+
+
+  //DBG LCH:check value  
+  // warm start, power consumption 600 nA
+    uint8_t sleep_state = 0x04;
+    write_opmode_command(RADIO_SET_SLEEP, &sleep_state, 1);
+  
+    _operation_mode = MODE_SLEEP;
+    ThisThread::sleep_for(2);
```

Comment:
2ms
Is it required to wait 2ms ?
in mbed-os the deep sleep mode can be entered / exited quite frequencly, as soon as there is nothing to do for 10ms. so 2ms is quite a lot, and any saving will end up in power saving (this comment is valid for all the timings that happen very frequencly)

Reply:
yes, this delay is present in sx126x radio Semtech implementation. 
Not sure how 2ms is calculated, because in sx126x datasheet, it's explained that  after SetSleep command, device will become unresponsive for around 0,5ms